### ISSUE TITLE
Compatibility of `cytoDArk0` Nissl-stained histological PNG patches with CellViT

### ISSUE THREAD

**Issue by shubhankar-git** [TIO-IKIM/CellViT#73]
Hello, 

Thank you for developing and sharing `CellViT`—it's an impressive tool for cell detection in histology images. I am interested in using it with the [cytoDArk0 dataset](https://zenodo.org/records/13694738), which consists of Nissl-stained histological images of the mammalian brain. However, this dataset only contains pre-extracted PNG patches rather than WSIs (Whole Slide Images).

Given that `CellViT`'s `cell_detection.py` script is designed primarily for WSIs, I wanted to check:

1. **Would it be feasible to use `cytoDArk0`'s PNG patches with `CellViT` for cell detection?**
   - I am aware that `CellViT` typically expects WSIs, from which patches are extracted in a preprocessing step. My dataset consists of pre-extracted patches, so I'm wondering if there’s a way to adapt `cell_detection.py` or if there is an example of custom dataloaders specifically for patch-based data.

2. **Is there a recommended approach for merging predictions across patches,** especially if there's overlap? Since some datasets, like `cytoDArk0`, provide patches with potential overlaps, any guidance on combining results from overlapping patches would be helpful.

3. **What modifications might be necessary to run inference on patch datasets** like `cytoDArk0`? If there are any example scripts or resources to guide custom inference, that would be much appreciated.

Thank you in advance for any insights, and I look forward to your guidance on using patch-only datasets with `CellViT`!

**Comment by shubhankar-git** [comment]
![image](https://github.com/user-attachments/assets/c2c4cd32-383a-4bce-921f-a19ae78cec3e)


I used the `cytoDArk` dataset, which consists of PNG patch images, with the `inference_cellvit_experiment_monuseg` code for inference. However, the results didn’t come out as expected.

Could you please advise on any specific changes or modifications I need to make in the inference code to ensure compatibility with patch-based datasets like `cytoDArk`? Any guidance on adapting the code or dataloader for non-WSI image patches would be greatly appreciated.

Thank you for your help! 




`
python3 ./cell_segmentation/inference/inference_cytodark.py \
  --model ./models/pretrained/CellViT/CellViT-SAM-H-x40.pth\
  --gpu 0 \
  --outdir  ./example/cytodark \
  --dataset /storage/cell_segmentation/cytoDArk0/20x/1024x1024/image \
  
`

**Comment by shubhankar-git** [comment]
![image](https://github.com/user-attachments/assets/ec41a33c-1490-458a-856d-fbc5aa5f8ff2)
After omitting patching, overlap, and other metrics calculations, and adjusting the dataloader, I obtained the `nuclei_binary_map` output.

Now, I’d like to know if there are any hyperparameters I can tune to improve the results.

**Comment by FabianHoerst** [comment]
Hi, the problem is that the magnifications might differ. Have you used x20 or x40 magnification as input patches?

**Comment by shubhankar-git** [comment]
I haven’t specifically applied magnification settings, as `calculate_step_metric_overlap` references them, but I’m not utilizing overlap since my PNG files seem to be pre-processed. This means I haven’t adjusted for magnification in the function either. Could you suggest any other hyperparameters or adjustments that might improve results?

**Comment by shubhankar-git** [comment]
I haven’t specifically applied magnification settings since `calculate_step_metric_overlap` references them, and I’m not doing overlap or doing evaluation. This means I haven’t adjusted for magnification in that function either. Are there other hyperparameters or adjustments you’d recommend to improve the results?

**Comment by shubhankar-git** [comment]
here is the code 

`def run_inference(self):
        """Run inference and save output images"""
        print('inference_running...')
        with torch.no_grad():
            for idx, (img, img_name) in enumerate(self.inference_dataloader):
                img = img.to(self.device)
              #  print(img)
                resize_transform = transforms.Resize((1024, 1024))

                # Apply the transform in your inference code
                img = resize_transform(img)

                # Run the model to get predictions
                predictions = self.model(img)
                print(predictions.keys)

                # Extract specific prediction (e.g., nuclei_binary_map)
                predicted_mask = predictions['nuclei_binary_map'].squeeze(0).cpu().numpy()
                # Select only the first channel if `nuclei_binary_map` contains multiple channels
                predicted_mask = predicted_mask[0]  # Take the first channel (adjust if needed)
                predicted_mask = (predicted_mask - predicted_mask.min()) / (predicted_mask.max() - predicted_mask.min())

                print(f"Predicted mask min: {predicted_mask.min()}, max: {predicted_mask.max()}")


                # Verify the shape to ensure it's now 2D
                print(f"Predicted mask shape after selecting a channel: {predicted_mask.shape}")
                # Ensure the file name has a clean .png extension
                img_name = img_name[0] if isinstance(img_name, (list, tuple)) else img_name  # Handle cases where img_name is in a list/tuple
                img_name = img_name.strip()  # Remove any whitespace or extra characters

                # Convert predicted_mask to an image and save
                output_img = Image.fromarray((predicted_mask * 255).astype(np.uint8))
                output_img.save(self.outdir / f"{img_name}")  # Explicitly add .png extension
                print(f"Saved {img_name}")
                break`

**Comment by shubhankar-git** [comment]
here is the code 

```
def run_inference(self):
        """Run inference and save output images"""
        print('inference_running...')
        with torch.no_grad():
            for idx, (img, img_name) in enumerate(self.inference_dataloader):
                img = img.to(self.device)
              #  print(img)
                resize_transform = transforms.Resize((1024, 1024))

                # Apply the transform in inference code
                img = resize_transform(img)

                # Run the model to get predictions
                predictions = self.model(img)
                print(predictions.keys)

                # Extract prediction
                predicted_mask = predictions['nuclei_binary_map'].squeeze(0).cpu().numpy()
                # Select only the first channel as  `nuclei_binary_map` contains multiple channels
                predicted_mask = predicted_mask[0]  
               #normalizing
                predicted_mask = (predicted_mask - predicted_mask.min()) / (predicted_mask.max() - predicted_mask.min())

                print(f"Predicted mask min: {predicted_mask.min()}, max: {predicted_mask.max()}")

                print(f"Predicted mask shape after selecting a channel: {predicted_mask.shape}")
          
                img_name = img_name[0] if isinstance(img_name, (list, tuple)) else img_name  
                img_name = img_name.strip()  # Remove any whitespace or extra characters

                # Convert predicted_mask to an image and save
                output_img = Image.fromarray((predicted_mask * 255).astype(np.uint8))
                output_img.save(self.outdir / f"{img_name}")  # Explicitly add .png extension
                print(f"Saved {img_name}")
                break```

**Comment by FabianHoerst** [comment]
Hi,

sorry for replying late. I currently have no capacity to go into this by detail to be honest. I think there are adaptions needed. I would recommend you to look into the existing codebase and make a fork for your dataset. I am very sorry that I cannot help you further in this matter..

Best!
